lib/Dialect/LLHD/Transforms/MemoryToBlockArgumentPass.cpp

```diff
@@ -0,0 +1,236 @@
+                         std::set<Block *> &closure) {
+  unsigned numElements;
+  for (Block *block : initialSet) {
+    getDominanceFrontier(block, op, closure);
+  }
+  do {
+    numElements = closure.size();
+    for (Block *block : closure) {
+      getDominanceFrontier(block, op, closure);
+    }
+  } while (numElements < closure.size());
+}
+
+/// Add a block argument to a given terminator. Only 'std.br', 'std.cond_br' and
+/// 'llhd.wait' are supported. The successor block has to be provided for the
+/// 'std.cond_br' terminator which has to possible successors.
+static void addBlockOperandToTerminator(Operation *terminator,
+                                        Block *successsor, Value toAppend) {
+  if (auto wait = dyn_cast<llhd::WaitOp>(terminator)) {
+    wait.destOpsMutable().append(toAppend);
+  } else if (auto br = dyn_cast<BranchOp>(terminator)) {
+    br.destOperandsMutable().append(toAppend);
+  } else if (auto condBr = dyn_cast<CondBranchOp>(terminator)) {
+    if (condBr.falseDest() == successsor) {
+      condBr.falseDestOperandsMutable().append(toAppend);
+    } else {
+      condBr.trueDestOperandsMutable().append(toAppend);
+    }
+  } else {
+    llvm_unreachable("unsupported terminator op");
+  }
+}
+
+void MemoryToBlockArgumentPass::runOnOperation() {
+  Operation *operation = getOperation();
+  OpBuilder builder(operation);
+
+  // No operations that have their own region and are not isolated from above
+  // are allowed for now.
+  WalkResult result = operation->walk([](Operation *op) -> WalkResult {
+    if (op->getNumRegions() > 0 && !op->isKnownIsolatedFromAbove())
+      return WalkResult::interrupt();
+    return WalkResult::advance();
+  });
+  if (result.wasInterrupted())
+    return;
+
+  // Get all variables defined in the body of this operation
+  // Note that variables that are passed as a function argument are not
+  // considered.
+  std::vector<Value> vars;
+  for (llhd::VarOp var : operation->getRegion(0).getOps<llhd::VarOp>()) {
+    vars.push_back(var.result());
+  }
+
+  // Don't consider variables that are used in other operations than load and
+  // store (e.g. as an argument to a call)
+  for (auto var = vars.begin(); var != vars.end(); ++var) {
+    for (Operation *user : var->getUsers()) {
+      if (!isa<llhd::LoadOp>(user) && !isa<llhd::StoreOp>(user)) {
+        vars.erase(var--);
+        break;
+      }
+    }
+  }
+
+  // For each variable find the blocks where a value is stored to it
+  for (Value var : vars) {
+    std::vector<Block *> defBlocks;
+    defBlocks.push_back(
+        var.getDefiningOp<llhd::VarOp>().getOperation()->getBlock());
+    operation->walk([&](llhd::StoreOp op) {
+      if (op.pointer() == var)
+        defBlocks.push_back(op.getOperation()->getBlock());
+    });
+    // Remove duplicates from the list
+    defBlocks.erase(std::unique(defBlocks.begin(), defBlocks.end()),
+                    defBlocks.end());
```

Comment:
`defBlocks` has two distinct phases: insert elements, then query. For this pattern a sorted vector is actually recommended by the [LLVM programmer manual](https://llvm.org/docs/ProgrammersManual.html#a-sorted-vector). This is because sorting and deleting duplicates after inserting all elements should be more efficient than checking for duplicates after every insertion. But yes, I indeed forgot to sort the vector first. Thanks!